### PR TITLE
Hail batch part1

### PR DESCRIPTION
**Pull request by annacuomo**
I think the first part of the hail batch pipeline (batch.py) is now ready to be reviewed, and would be useful to start running / debugging this on main, thanks!

### CODE REVIEW

File: batch.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
+    mt_path = output_path("in_window.mt", "tmp")
```
Comment (MattWellie) [inline batch.py]:
```suggestion
    mt_path = output_path(f"{gene_name}_in_window.mt", "tmp")
```

Interested in knowing if this is a useful checkpoint - the only filtering done up to this point is on a table key (Chr-Pos), so I would expect hail to execute that without much computational overhead. It's good to remember that checkpointing isn't 'free', but I don't have a good idea of how much (if at all) it will speed up the annotation step

File: batch.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    promoter_path = output_path(f"{gene_name}promoter_variants.mt", "tmp")
+    mt = mt.checkpoint(promoter_path, overwrite=True)  # checkpoint
+    logging.info(
+        f"Number of rare (freq<5%) QC-passing, biallelic SNPs in promoter regions: {mt.count()[0]}"
+    )
+
+    # export this as a Hail table for downstream analysis
+    ht_path = output_path(f"{gene_name}_rare_promoter_summary.ht")
+    ht = mt.rows()
+    ht.write(ht_path)
```
Comment (MattWellie) [inline batch.py:188]: Here you're checkpointing the MT, writing the MT as plink, and also writing the rows. Hail can export to plink, and import from plink. Could you replace these 3 writes with just the write to plink? (then [import_plink](https://hail.is/docs/0.2/methods/impex.html#hail.methods.import_plink) later if you need an alternate format?)

File: batch.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    mt = hl.read_matrix_table(mt_path)
+
+    gene_name = gene_details["gene_name"]
+
+    # get relevant chromosome
+    chrom = gene_details["chrom"]
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    left_boundary = max(1, int(gene_details["gene-start"]) - window_size)
```
Comment (MattWellie) [inline batch.py]: I thought in your demo file the columns were `start` and `end` 👀

File: batch.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        plink_job.call(
+            get_promoter_variants,
+            mt_path=output_mt_path,
+            ht_path=anno_ht_path,
+            gene_details=gene_dict[gene],
+            window_size=window_size,
+            plink_file=plink_file,
+        )
+        # concatenate jobs so they can be depended on
+        genotype_jobs.append(plink_job)
```
Comment (MattWellie) [inline batch.py]: You'll need to do a `batch.run(wait=False)` to set all the jobs running

File: batch.py (hunk 3 above)
Comment (annacuomo) [reply batch.py]: yes you are right, i had already misunderstood how this worked and i had renames the columns (i thought specifying it was the gene's start, vs e.g. the variant was useful) - but with the current syntax this is wrong, you're right!

File: batch.py (hunk 4 above)
Comment (annacuomo) [reply batch.py]: nice yea that makes sense

File: batch.py (hunk 2 above)
Comment (annacuomo) [reply batch.py:188]: do you mean specifically in case i won't actually need the hail table? because i definitely will want to save those records. Also does the export to plink definitely retain all information?

Comment (MattWellie) [reply batch.py:188]: That's a good question 😅 I only saw that an import-from plink method exists, but I don't know if the MT -> Plink -> MT process is lossy.

My main thought is that you're checkpointing, then writing two objects - the checkpoint is a good idea to make both other writes faster, but if you don't explicitly need the hail data you can reduce this down to just the one write to disc (plink). I'm not really familiar with the plink output format, so it might drop all the annotations you need (in which case the hail table is probably good to write, and this section is fine)

Comment (annacuomo) [reply batch.py:188]: yea that was my assumption, so may leave it for now (I don't think the plink format is well suited for all these annotations, though I don't really know)

File: batch.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    # submit a job for each gene (export genotypes to plink)
+    genotype_jobs = []
+    for gene in genes_of_interest:
+
+        # final path for this gene - generate first (check syntax)
+        plink_file = output_path(f"{output_path}/plink_files/{gene}")
+        gene_dict[gene]["plink"] = plink_file
+
+        # if the plink output exists, do not re-generate it
+        if to_path(plink_file).exists():
```
Comment (MattWellie) [inline batch.py]: I'd like to check my understanding here - this file path is being generated as a prefix, and `export_plink` will create a few files (fam/bim/etc)?

If that's the case, this prefix will never exist, and you'll need to check for the existence of a specific file as a proxy for 'was the plink data generated'

```suggestion
        if to_path(f'{plink_file}.bim').exists():
```

Comment (annacuomo) [reply batch.py]: yup good shout

File: batch.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+@click.option("--output-path")
+@click.option("--mt-path")
+@click.option("--anno-ht-path")
+def crm_pipeline(
+    chromosomes: list[str],
+    genes: list[str],
+    celltypes: list[str],
+    expression_files_prefix: str,  # 'scrna-seq/grch38_association_files'
+    sample_mapping_file: str,
+    output_path: str,  # 'tob_wgs_rv/pseudobulk_rv_association'
```
Comment (MattWellie) [inline batch.py]: Calling this variable output_path will obscure the `output_path` method, you should choose a different name!

File: batch.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    """
+    outliers = ["966_967", "88_88"]
+    df = df[-df.sampleid.isin(outliers)]
+
+    return df
+
+
+# endregion MISCELLANEOUS
+
+config = get_config()
```
Comment (MattWellie) [inline batch.py]:
```suggestion
```

File: batch.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    expression_files_prefix: str,  # 'scrna-seq/grch38_association_files'
+    sample_mapping_file: str,
+    output_path: str,  # 'tob_wgs_rv/pseudobulk_rv_association'
+    mt_path: str = DEFAULT_JOINT_CALL_MT,  # 'mt/v7.mt'
+    anno_ht_path: str = DEFAULT_ANNOTATION_HT,  # 'tob_wgs_vep/104/vep104.3_GRCh38.ht'
+    window_size: int = 50000,
+):
+
+    sb = hb.ServiceBackend(
+        billing_project=config["hail"]["billing_project"],
```
Comment (MattWellie) [inline batch.py]:
```suggestion
        billing_project=get_config()["hail"]["billing_project"],
```

**Comment by MattWellie** [comment]
Another change to make your life easier throughout this script - try using `get_batch()` ([example](https://github.com/populationgenomics/automated-interpretation-pipeline/blob/main/reanalysis/interpretation_runner.py#L37))

This means that you don't have to explicitly pass a reference to a batch around your code in order to add jobs to it. The actual method is [defined here](https://github.com/populationgenomics/production-pipelines/blob/main/cpg_workflows/batch.py#L25) if you want a look. Using it is very similar to the `get_config()` syntax.

File: batch.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+)  # consider removing some print statements?
+
+DEFAULT_JOINT_CALL_MT = dataset_path("mt/v7.mt")
+DEFAULT_ANNOTATION_HT = dataset_path(
+    "tob_wgs_vep/104/vep104.3_GRCh38.ht"
+)  # atm VEP only
+
+CELLREGMAP_IMAGE = (
+    "australia-southeast1-docker.pkg.dev/cpg-common/images/cellregmap:0.0.3"
+)
```
Comment (lgruen) [inline batch.py]: Is the always the same as the driver image that's passed to the `analysis-runner` invocation? If so, you can fetch this value from the config: `get_config()['workflow']['driver_image']`.

File: batch.py [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+    output_mt_path: str,  # "tob_wgs_rv/densified_rv_only.mt"
+):
+    """Subset hail matrix table
+
+    Input:
+    joint call hail matrix table
+    set of samples for which we have scRNA-seq data
+
+    Output:
+    subsetted hail matrix table, containing only variants that:
```
Comment (lgruen) [inline batch.py]:
```suggestion
    subset hail matrix table, containing only variants that:
```

File: batch.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        mt_path=mt_path,
+        samples=sc_samples,
+        output_mt_path=output_mt_path,
+    )
+
+    # grab all relevant genes across all chromosomes
+    # simpler if gene details are condensed to one file
+    gene_dict: dict[str, dict] = {}
+    for chromosome in chromosomes:
+        geneloc_tsv_path = os.path.join(
```
Comment (lgruen) [inline batch.py:282]: Could you fetch the gene intervals directly from GENCODE instead and just pass a list of gene IDs?

File: tasks/R_task.wdl [hunk 12]
```diff
@@ -0,0 +1,21 @@
+task R_Task {
```
Comment (lgruen) [inline tasks/R_task.wdl]: This file and the R Dockerfile seems unrelated?

File: batch.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+        plink_job.call(
+            get_promoter_variants,
+            mt_path=output_mt_path,
+            ht_path=anno_ht_path,
+            gene_details=gene_dict[gene],
+            window_size=window_size,
+            plink_file=plink_file,
+        )
+        # concatenate jobs so they can be depended on
+        genotype_jobs.append(plink_job)
```
Comment (lgruen) [inline batch.py]: You're not using this currently, so could be removed?

File: batch.py [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+        # final path for this gene - generate first (check syntax)
+        plink_file = output_path(f"{output_path}/plink_files/{gene}")
+        gene_dict[gene]["plink"] = plink_file
+
+        # if the plink output exists, do not re-generate it
+        if to_path(f'{plink_file}.bim').exists():
+            continue
+
+        plink_job = batch.new_python_job(f"Create plink files for: {gene}")
```
Comment (lgruen) [inline batch.py:306]: Maybe a tangent, but I thought worth at least asking: are you sure this needs to be done in PLINK or could we use Hail instead? Then you wouldn't need to worry about scattering across genes etc.

File: batch.py [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+@click.option("--expression-file-prefix")
+@click.option("--sample-mapping-file")
+@click.option("--output-path")
+@click.option("--mt-path")
+@click.option("--anno-ht-path")
+def crm_pipeline(
+    chromosomes: list[str],
+    genes: list[str],
+    celltypes: list[str],
+    expression_files_prefix: str,  # 'scrna-seq/grch38_association_files'
```
Comment (lgruen) [inline batch.py]: Generally it's better to add such comments to the `help` or `default` field of the corresponding `click.option`.

File: batch.py [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt = mt.checkpoint(output_mt_path, overwrite=True)
```
Comment (lgruen) [inline batch.py]: You should only use checkpoint for temporary results, as `checkpoint` uses a faster codec that produces larger results. For permanent results, use `write` instead.

You'll have to reread the matrix table again afterwards though, as `write` doesn't return anything.

File: batch.py [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+    return gene_dict
+
+
+# copied from https://github.com/populationgenomics/tob-wgs/blob/main/scripts/eqtl_hail_batch/launch_eqtl_spearman.py
+# check whether it needs modifying
+def remove_sc_outliers(df):
+    """
+    Remove outlier samples, as identified by sc analysis
+    """
+    outliers = ["966_967", "88_88"]
```
Comment (lgruen) [inline batch.py]: Should this be a parameter instead of hardcoding these values?

**Review by lgruen — APPROVED** [review]
Please address the review comments before merging, but overall looks fine!

File: batch.py (hunk 9 above)
Comment (annacuomo) [reply batch.py]: yes it should be, thanks good to know!

File: batch.py (hunk 6 above)
Comment (annacuomo) [reply batch.py]: yes of course, thank you!

File: batch.py (hunk 9 above)
Comment (annacuomo) [reply batch.py]: i have tried but am slightly unsure of the syntax?

File: batch.py (hunk 11 above)
Comment (annacuomo) [reply batch.py:282]: i presume the file(s) i am using here do come from gencode already, but I am not sure what you mean exactly?

File: tasks/R_task.wdl (hunk 12 above)
Comment (annacuomo) [reply tasks/R_task.wdl]: i have removed from this PR

File: batch.py (hunk 13 above)
Comment (annacuomo) [reply batch.py]: yes good point!

File: batch.py (hunk 14 above)
Comment (annacuomo) [reply batch.py:306]: I cannot remember whether we already had this conversation, but I could not figure out a handy solution to export (from a hail matrix table) a matrix of samples by variants with the genotype calls inside? If there was, that would be amazing!

File: batch.py (hunk 15 above)
Comment (annacuomo) [reply batch.py]: ok, i should've done?

File: batch.py (hunk 17 above)
Comment (annacuomo) [reply batch.py]: yes good point

File: batch.py (hunk 16 above)
Comment (annacuomo) [reply batch.py]: i think in this case i was not returning anything anyways, so should be ok to just turn this checkpoint into write?

File: batch.py (hunk 11 above)
Comment (MattWellie) [reply batch.py:282]: @lgruen do you mean polling the [Ensembl REST API](https://rest.ensembl.org/documentation/info/symbol_lookup), for example?

```https://rest.ensembl.org/lookup/symbol/homo_sapiens/BRCA2?content-type=application/json```

**Review by MattWellie — APPROVED** [review]
